core/src/main/java/org/apache/calcite/sql/validate/SqlValidatorImpl.java

```diff
@@ -1217,12 +1219,21 @@ protected void validateNamespace(final SqlValidatorNamespace namespace,
         // fields; these are neutralized if the consuming query filters on them.
         final ImmutableBitSet mustFilterFields =
             namespace.getMustFilterFields();
-        if (!mustFilterFields.isEmpty()) {
+        // Remnant must filter fields are fields that are not selected and cannot
```

Comment:
Not sure whether the 'remnant' concept is necessary. Instead could we just remove must-filter fields that have been satisfied?

Reply:
I think it is necessary to handle the case where the bypass field is selected, but the must-filter fields are not selected. 

Imagine this setup
```
Table: EMP (empno, name, id)
must_filters: empno
bypass_filter: name
```
Query:
`select * from (select name from emp) where name = 'bob' `

The inner query `select name from emp` can't error on `empno` not being filtered at this point because `name` was selected, and could be filtered on later. The must-filter for `empno` is not satisfied yet at this point. 

When we get to the `select * from _ where name = 'bob'`, that's where we defuse the error for `empno` based on seeing `name` being filtered on. 


If alternatively the query was 
`select * from (select name, id from emp) where id = '12' `
When we get to `select * from _ where id = '12'`, we're at the top-level, and remnantMustFilter's set is not empty, hence we throw an error.

Reply:
I see. If `t` has a must-filter field `f` and bypass fields `b0` and `b1`, then the query `select b0, b1 from t` has no must-filter fields, no bypass fields, but two remnant fields `[b0, b1]`.

If you filter on any of those remnant fields then they all go away.

Reply:
Actually that is not how the sets are populated.
The purpose of remnant fields is to capture the fields that were marked as required, but no longer accessible to be filtered on and defused in the select clause. In other words, you cannot filter on any of the remnant fields. The presence of values being in the remant-filter set serves the purpose of being an indicator for erroring. 

They can only be defused if any of the bypass fields are filtered on.

In your example this would be the correct set values:
`select b0, b1 from t`

must-filter field: []
bypass-fields: [`b0`, `b1`]
remnant-fields: [f]